File: packages/native/ios/JigHandshakeHandler.m
#import "JigHandshakeHandler.h"

@interface JigHandshakeHandler ()
@property (nonatomic, copy) NSString *rnVersion;
@property (nonatomic, copy, nullable) NSString *expoVersion;
@property (nonatomic, copy) NSString *appName;
@property (nonatomic, copy) NSString *bundleId;
@end

@implementation JigHandshakeHandler

+ (NSInteger)protocolVersion { return 1; }
+ (NSString *)serverVersion { return @"0.1.0"; }

- (instancetype)initWithRnVersion:(NSString *)rnVersion
                     expoVersion:(NSString *)expoVersion {
    self = [super init];
    if (self) {
        _rnVersion = [rnVersion copy];
        _expoVersion = [expoVersion copy];
        _appName = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleName"] ?: @"Unknown";
        _bundleId = [NSBundle mainBundle].bundleIdentifier ?: @"unknown";
    }
    return self;
}

- (NSString *)makeServerHello {
    NSMutableDictionary *appDict = [@{
        @"name": self.appName,
        @"bundleId": self.bundleId,
        @"platform": @"ios",
        @"rn": self.rnVersion,
    } mutableCopy];
    if (self.expoVersion) {
        appDict[@"expo"] = self.expoVersion;
    }

    NSDictionary *params = @{
        @"protocol": @{
            @"name": @"jig",
            @"version": @([self.class protocolVersion]),
            @"min": @([self.class protocolVersion]),
            @"max": @([self.class protocolVersion]),
        },
        @"server": [self.class serverVersion],
        @"app": [appDict copy],
        @"commands": @[],
        @"domains": @[],
    };

    NSDictionary *message = @{
        @"jsonrpc": @"2.0",
        @"method": @"server.hello",
        @"params": params,
    };

    return [self jsonStringFromDictionary:message];
}

- (nullable NSString *)handleMessage:(NSString *)text {
    NSData *data = [text dataUsingEncoding:NSUTF8StringEncoding];
    if (!data) return nil;

    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
    if (![json isKindOfClass:[NSDictionary class]]) return nil;

    NSString *method = json[@"method"];
    if ([method isEqualToString:@"client.hello"]) {
        return [self handleClientHello:json];
    }

    return nil;
}

#pragma mark - Private

- (NSString *)handleClientHello:(NSDictionary *)json {
    NSInteger msgId = [json[@"id"] integerValue];
    NSDictionary *params = json[@"params"];
    NSDictionary *clientProtocol = params[@"protocol"];
    NSInteger requestedVersion = [clientProtocol[@"version"] integerValue];

    if (!params || !clientProtocol || requestedVersion == 0) {
        return [self makeErrorWithId:msgId code:-32600 message:@"Invalid client.hello params" data:nil];
    }

    NSInteger ours = [self.class protocolVersion];
    if (requestedVersion < ours || requestedVersion > ours) {
        NSDictionary *errorData = @{
            @"supported": @{ @"min": @(ours), @"max": @(ours) }
        };
        NSString *msg = [NSString stringWithFormat:
            @"Protocol version %ld not supported. Server supports version %ld.",
            (long)requestedVersion, (long)ours];
        return [self makeErrorWithId:msgId code:-32001 message:msg data:errorData];
    }

    NSString *sessionId = [NSString stringWithFormat:@"sess_%@",
        [[[NSUUID UUID] UUIDString] substringToIndex:8].lowercaseString];

    NSDictionary *result = @{
        @"sessionId": sessionId,
        @"negotiated": @{ @"protocol": @(requestedVersion) },
        @"enabled": @[],
        @"rejected": @[],
    };

    NSDictionary *response = @{
        @"jsonrpc": @"2.0",
        @"id": @(msgId),
        @"result": result,
    };

    return [self jsonStringFromDictionary:response];
}

- (NSString *)makeErrorWithId:(NSInteger)msgId
                         code:(NSInteger)code
                      message:(NSString *)message
                         data:(nullable NSDictionary *)data {
    NSMutableDictionary *error = [@{
        @"code": @(code),
        @"message": message,
    } mutableCopy];
    if (data) {
        error[@"data"] = data;
    }

    NSDictionary *response = @{
        @"jsonrpc": @"2.0",
        @"id": @(msgId),
        @"error": [error copy],
    };

    return [self jsonStringFromDictionary:response];
}

- (NSString *)jsonStringFromDictionary:(NSDictionary *)dict {
    NSData *data = [NSJSONSerialization dataWithJSONObject:dict
                                                  options:NSJSONWritingSortedKeys
                                                    error:nil];
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

@end

File: packages/native/android/src/main/java/expo/modules/jig/JigHandshakeHandler.java
package expo.modules.jig;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.UUID;

public class JigHandshakeHandler {
    public static final int PROTOCOL_VERSION = 1;
    public static final String SERVER_VERSION = "0.1.0";

    private final String appName;
    private final String packageName;
    private final String rnVersion;
    private final String expoVersion; // nullable

    public JigHandshakeHandler(String appName, String packageName,
                               String rnVersion, String expoVersion) {
        this.appName = appName;
        this.packageName = packageName;
        this.rnVersion = rnVersion;
        this.expoVersion = expoVersion;
    }

    public String makeServerHello() {
        try {
            JSONObject app = new JSONObject();
            app.put("name", appName);
            app.put("bundleId", packageName);
            app.put("platform", "android");
            app.put("rn", rnVersion);
            if (expoVersion != null) {
                app.put("expo", expoVersion);
            }

            JSONObject protocol = new JSONObject();
            protocol.put("name", "jig");
            protocol.put("version", PROTOCOL_VERSION);
            protocol.put("min", PROTOCOL_VERSION);
            protocol.put("max", PROTOCOL_VERSION);

            JSONObject params = new JSONObject();
            params.put("protocol", protocol);
            params.put("server", SERVER_VERSION);
            params.put("app", app);
            params.put("commands", new JSONArray());
            params.put("domains", new JSONArray());

            JSONObject message = new JSONObject();
            message.put("jsonrpc", "2.0");
            message.put("method", "server.hello");
            message.put("params", params);

            return message.toString();
        } catch (JSONException e) {
            throw new RuntimeException("Failed to build server.hello", e);
        }
    }

    public String handleMessage(String text) {
        try {
            JSONObject json = new JSONObject(text);
            String method = json.optString("method", "");

            if ("client.hello".equals(method)) {
                return handleClientHello(json);
            }
        } catch (JSONException e) {
            // Malformed JSON — ignore
        }
        return null;
    }

    private String handleClientHello(JSONObject json) throws JSONException {
        int id = json.optInt("id", 0);
        JSONObject params = json.optJSONObject("params");
        if (params == null) {
            return makeError(id, -32600, "Invalid client.hello params", null);
        }

        JSONObject clientProtocol = params.optJSONObject("protocol");
        if (clientProtocol == null) {
            return makeError(id, -32600, "Invalid client.hello params", null);
        }

        int requestedVersion = clientProtocol.optInt("version", 0);
        if (requestedVersion < PROTOCOL_VERSION || requestedVersion > PROTOCOL_VERSION) {
            JSONObject supported = new JSONObject();
            supported.put("min", PROTOCOL_VERSION);
            supported.put("max", PROTOCOL_VERSION);
            JSONObject errorData = new JSONObject();
            errorData.put("supported", supported);

            return makeError(id, -32001,
                "Protocol version " + requestedVersion +
                " not supported. Server supports version " + PROTOCOL_VERSION + ".",
                errorData);
        }

        String sessionId = "sess_" + UUID.randomUUID().toString().substring(0, 8);

        JSONObject negotiated = new JSONObject();
        negotiated.put("protocol", requestedVersion);

        JSONObject result = new JSONObject();
        result.put("sessionId", sessionId);
        result.put("negotiated", negotiated);
        result.put("enabled", new JSONArray());
        result.put("rejected", new JSONArray());

        JSONObject response = new JSONObject();
        response.put("jsonrpc", "2.0");
        response.put("id", id);
        response.put("result", result);

        return response.toString();
    }

    private String makeError(int id, int code, String message, JSONObject data) throws JSONException {
        JSONObject error = new JSONObject();
        error.put("code", code);
        error.put("message", message);
        if (data != null) {
            error.put("data", data);
        }

        JSONObject response = new JSONObject();
        response.put("jsonrpc", "2.0");
        response.put("id", id);
        response.put("error", error);

        return response.toString();
    }
}
